kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java

```diff
@@ -779,22 +793,42 @@ public void findEntryComplete(Position position, Object ctx) {
                         }
 
+
                         if (log.isDebugEnabled()) {
                             log.debug("Find position for topic {} time {}. position: {}",
                                 perTopic.getName(), timestamp, finalPosition);
                         }
-                        if (legacyMode) {
-                            partitionData.complete(new ListOffsetResponse.PartitionData(
-                                    Errors.NONE,
-                                    Collections.singletonList(
-                                            MessageIdUtils.getOffset(
-                                                    finalPosition.getLedgerId(), finalPosition.getEntryId()))));
-                        } else {
-                            partitionData.complete(new ListOffsetResponse.PartitionData(
-                                    Errors.NONE,
-                                    DEFAULT_TIMESTAMP,
-                                    MessageIdUtils.getOffset(finalPosition.getLedgerId(), finalPosition.getEntryId())));
-                        }
+
+                        MessageIdUtils.getOffsetOfPosition(managedLedger, finalPosition)
+                                .whenComplete((offset, throwable) -> {
+                            if (throwable != null) {
+                                log.error("[{}] Failed to get offset for position {}",
+                                        perTopic, finalPosition, throwable);
+                                if (legacyMode) {
+                                    partitionData.complete(new ListOffsetResponse
+                                            .PartitionData(
+                                            Errors.UNKNOWN_SERVER_ERROR,
+                                            Collections.emptyList()));
+                                } else {
+                                    partitionData.complete(new ListOffsetResponse
+                                            .PartitionData(
+                                            Errors.UNKNOWN_SERVER_ERROR,
+                                            ListOffsetResponse.UNKNOWN_TIMESTAMP,
+                                            ListOffsetResponse.UNKNOWN_OFFSET));
+                                }
+                                return;
+                            }
+                            if (legacyMode) {
+                                partitionData.complete(new ListOffsetResponse.PartitionData(
+                                        Errors.NONE,
+                                        Collections.singletonList(offset)));
+                            } else {
+                                partitionData.complete(new ListOffsetResponse.PartitionData(
+                                        Errors.NONE,
+                                        DEFAULT_TIMESTAMP,
+                                        offset));
+                            }
+                        });
```

Comment:
Should we add a method to do the `get offset -> fill PartitionData` logic since these two changes of `fetchOffsetForTimestamp` are totally same?

Reply:
I'll add a new method for this logic.